### ISSUE TITLE
Phone needs to connect first after every power cycle

### ISSUE THREAD

**Issue by kipwittchen** [jim-mckeown/SP200A-Client#1]
Hello,

I successfully have the script running! The only issue I'm running into is when I first turn the SP200A on, if the esp32 is the first device to connect to its wifi, data doesn't seem to flow in, serial just keeps outputting Depth: --.-- and if I try connecting with the phone, it displays the error "No SonarPhone WiFi Signal Found". If I then turn off the SP200A and esp32, then only turn the SP200A on and connect to it via phone, it works, and then turn on esp32 and it connects pulling in depth data normally. Phone at this point can disconnect and everything keeps working. The red light on the SP200A also will stay solid under the first scenario and will do the normal slow flash under the second working scenario.

Here's my script at the moment, I've cleared out the seatalk protocol parts because I won't be using that, maybe I cleared too much out? 

```
#include "WiFi.h"
#include "AsyncUDP.h"
#include <WiFiManager.h>

const char * ssid = "T-BOX-720";
const char * password = "";
uint8_t FC[] = {70,67,21,0,244,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0}; // mac address, units, beam width, depth max/min, checksum not set
uint8_t FX[] = {70,88,21,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,179,0,0,0,0,0,0,0,0,0}; // 

float metersToFeet = 3.28084;

bool haveMac = false;
int depthMin = 0;
int depthMax = 0; 
int beamWidth = 0; // 8 = 20 deg, 2 = 40 deg
int dataCount = 0;
int depthUnits = 0; //0 = meters, 1 = feet
float depth = 0.0;
float depthFrac = 0.0;
float vBatt = 0.0;
float vBattFrac = 0.0;
float temp = 0.0;
int loopCount = 0;
bool newData = false;
bool FCsent = false;

#define RX1 2
#define TX1 4

AsyncUDP udp; 
 
bool sendSonarData()
{
  float depthToSend = 0.0;

  // set depth units to feet - convert meters to feet
  if(depthUnits == 1)
  {
    depthToSend = depth; 
  }
  else
  {
    depthToSend = depth * metersToFeet;
  }
  
  \\other code here to send via Serial1, removed right now for testing
  return true;
}

void setup()
{
    Serial.begin(115200);
    Serial1.begin(115200,SERIAL_8N1,RX1,TX1); //for UART to other arduino
  
    //WiFiManager intialization
    WiFi.mode(WIFI_STA);
    WiFiManager wm;
    //wm.resetSettings();
    bool res;
    res = wm.autoConnect("AutoConnectAP"); // anonymous ap
    if(!res) 
    {
        Serial.println("Failed to connect");
        // ESP.restart();
    } 
    else 
    {
        Serial.println("Connected to WiFi");
    }    
    
    if(udp.connect(IPAddress(192,168,1,1), 5000)) 
    {
        Serial.println("UDP connected");
        //udp.write(FC, 29);
        udp.onPacket([](AsyncUDPPacket packet) {
            // check for REDYFX. Update FC request
            if(packet.data()[6] == 82 && packet.data()[7] == 69 && packet.data()[8] == 68 && packet.data()[9] == 89 && packet.data()[10] == 70 && packet.data()[11] == 88) // is this REDYFX?
            {
              // set FC mac address
              for(int i = 0;i < 6;i++)
              {
                FC[21 + i] = packet.data()[26 + i];
              }
              
              // set units
              FC[11] = depthUnits;              
              
              // set max, min depth (4 bytes)
              FC[6] = depthMin % 256; // lower byte
              FC[7] = depthMin / 256; // upper byte
              FC[8] = depthMax % 256; // lower byte
              FC[9] = depthMax / 256; // upper byte
              
              // set beam width
              FC[13] = beamWidth;
              
              // set checksum
              int checksum = 0;
              for(int i = 0;i < 19;i++)
              {
                checksum += FC[i];
              }
              FC[19] = checksum % 256; // lower byte
              FC[20] = checksum / 256; // upper byte
              haveMac = true;
            }
            // check for REDYFC. Set Units, Depth, Range, Temp, BatteryVolts, newData flag
            if(packet.data()[6] == 82 && packet.data()[7] == 69 && packet.data()[8] == 68 && packet.data()[9] == 89 && packet.data()[10] == 70 && packet.data()[11] == 67) // is this REDYFC?
            {
              depthUnits = packet.data()[21];
              depthFrac = packet.data()[25];
              depth = packet.data()[23] + (packet.data()[24] * 256) + (depthFrac / 100);
              depthMin = packet.data()[16] + (packet.data()[17] * 256);
              depthMax = packet.data()[18] + (packet.data()[19] * 256);
              beamWidth = packet.data()[32];
              vBattFrac = packet.data()[31];
              vBatt = packet.data()[30] + (vBattFrac / 100);
              temp = packet.data()[26] * 9.0/5.0 + 32.0;
              newData = true;
            }
            dataCount++;
            packet.flush();          
        });
        delay(1000);
    }

    
}

void loop()
{
    if(haveMac)
    {
      if(newData)
      {
        Serial.printf("Depth: %.1f ", depth);
        //Serial.println();
        //Serial.printf("Temp: %.1f", temp);
        //Serial.println();
        //Serial.printf("vBatt: %.1f ", vBatt);
        if(!sendSonarData())
        {
          Serial.println("Error in sendSonarData()");
        }
        newData = false;
      }
      else
      {
        Serial.print("Depth: --.-- ");
      }
      
      
      if(FCsent)
      {
        Serial.println("-");
        FCsent = false;
      }
      else
      {
        Serial.println("");
      }
      
      if(loopCount > 19)
      {
        // update units
        FC[11] = depthUnits;
        
        // update max, min depth (4 bytes)
        FC[6] = depthMin % 256; // lower byte
        FC[7] = depthMin / 256; // upper byte
        FC[8] = depthMax % 256; // lower byte
        FC[9] = depthMax / 256; // upper byte
  
        // set beam width
        FC[13] = beamWidth;
        
        // update checksum
        int checksum = 0;
        for(int i = 0;i < 19;i++)
        {
          checksum += FC[i];
        }
        FC[19] = checksum % 256; // lower byte
        FC[20] = checksum / 256; // upper byte      
        
        dataCount = 0;
        udp.write(FC, sizeof FC);
        loopCount = 0;
        FCsent = true;
     }
      
      loopCount++;
      delay(500);
    }
    else
    {
      udp.write(FX, sizeof FX);
      delay(1000);
    }
}
```

**Comment by jim-mckeown** [comment]
Hello, and thanks for your feedback.

With your modified code, how long does the ESP32 continue to receive data from the SP200A after the phone app is disconnected? If the data stops after 40 to 60 seconds, that would be consistent with the ESP32 acting as a slave. The SP200A requires a valid FC request at about 30 second intervals to keep the REDYFC data going. The ESP32 may not be constructing a valid FC request. 

If you flash the original code, does the ESP32 operate as a single master? It should work as a master even without any SeaTalk hardware connected. 

- Jim

**Comment by kipwittchen** [comment]
Hi Jim,

Here's what I did to test this:
Factory reset reset SP200A, connected to SPA200A wifi with my phone (device MAC turned on, not randomized), signed in as master with default password. wiped flash on esp32 and flashed you original code. Connected to AP (device MAC turned on, not randomized), provided wifi credentials to SP200A (all defaults). From there the esp32 worked. I then turned off the SP200A, esp32, and put my phone in airplane mode. I turned on the SP200A, waited a minute and then turned on esp32, no data was received, depth reported as --.--. Connected to SP200A wifi via phone but the app couldn't connect either. Turned off/on SP200A and then the phone app could connect, restarted esp32 and it received depth data correctly. Put phone back in airplane mode and the esp32 kept receiving depth data after 30 seconds, continued working over the next few minutes I tested. Also saw the extra - for FCsent. Again though, turning everything off cause the esp32 to not connect correctly first.

Do I need to update the FC declaration with my phone's mac address? Or is that supposed to be stored to the esp32's memory? I see the line `//udp.write(FC, 29);` commented out, is that not supposed to be?

Thanks,
Kip

**Comment by kipwittchen** [comment]
Figured it out! I missed that `beamWidth` needed to be set to something other than 0. That must have caused it to error out until restarted. I set it to 8 (and units to 1 for feet) and now it's working perfectly every startup!

Edit: Also here's my fork of the project which communicates to an esphome device via uart to get the data into home assistant: https://github.com/kipwittchen/SP200A-Client-esphome-UART
